### DISCUSSION TITLE
MIME_BAD_UNICODE on screenshots coming from MacOS (?)

### DISCUSSION THREAD

**Discussion by ManChicken1911** [rspamd/rspamd#4893] [Q&A]
I have a couple of clients who I am pretty sure are on Macs of some sort and often send us attachments containing screenshots. The files are called "2024-03-26 at 12.16.15 PM.png" where the space between the **15** and **PM** is actually a unicode 'narrow non-breaking space' (0x202F) because, for whatever reason, someone at Apple thinks we need to have that in a filename instead of just a plain space.

Anyway, I can add a rule to one of my existing multimaps with a negative score to try and offset them, but I'm bringing it up as an open question if there should maybe be some sort of exception to filenames that match this exact format to the MIME_BAD_UNICODE?

**Comment by moisseev** [comment]
Nothing is wrong with this file name format. A filename that contains Unicode characters should be encoded and it is likely not in this case. So the problem is a broken MUA. I'd suggest to lower MIME_BAD_UNICODE symbol score for the particular version(s) of MUA.

**Reply by ManChicken1911** [reply]
It seems like it is encoded. The start of the MIME part on one the attachments from the raw message (base64 wrapped in a quoted-printable?)

```
--_005_DU0PR02MB958686DDB16B6147CABD868C85352DU0PR02MB9586eurp_
Content-Type: image/png;
    name="=?utf-8?B?U2NyZWVuc2hvdCAyMDI0LTAzLTI2IGF0IDEyLjE2LjE14oCvUE0ucG5n?="
Content-Description:
 =?utf-8?B?U2NyZWVuc2hvdCAyMDI0LTAzLTI2IGF0IDEyLjE2LjE14oCvUE0ucG5n?=
Content-Disposition: inline;
    filename="=?utf-8?B?U2NyZWVuc2hvdCAyMDI0LTAzLTI2IGF0IDEyLjE2LjE14oCvUE0ucG5n?=";
    size=212214; creation-date="Tue, 26 Mar 2024 18:18:21 GMT";
    modification-date="Tue, 26 Mar 2024 20:21:55 GMT"
Content-ID: <8a1a4f26-1a1b-44d7-9d96-5a5705f6a4f9>
Content-Transfer-Encoding: base64
```
The message headers don't reveal the MUA but it's coming from some hosted Exchange server so I'm assuming Outlook (and Outlook is perfect, right?! 🤣) So I took the saved file with the NNBSP character in the name, attached it to a new message to myself using Thunderbird, and rspamd sees it as MIME_BAD_UNICODE also - Thunderbird encoded it as such:

```
Content-Type: image/png;
 name="=?UTF-8?Q?Screenshot_2024-03-26_at_12=2E19=2E12=E2=80=AFPM=2Epng?="
Content-Disposition: attachment;
 filename*0*=UTF-8''%53%63%72%65%65%6E%73%68%6F%74%20%32%30%32%34%2D%30%33;
 filename*1*=%2D%32%36%20%61%74%20%31%32%2E%31%39%2E%31%32%E2%80%AF%50%4D;
 filename*2*=%2E%70%6E%67
Content-Transfer-Encoding: base64
```

**Comment by moisseev** [comment]
You are right. I made a wrong assumption. The MIME_BAD_UNICODE is specifically designed to catch such things. [util.has_obscured_unicode(str)](https://rspamd.com/doc/lua/rspamd_util.html#f53782)

Looking through my notes I found some Kyocera scanners may insert `0x200ed` in filenames.

I even had a composite for it:
```
LOCAL_MIME_BAD_UNICODE_KYOSCAN {
    expression = "MIME_BAD_UNICODE[/^0x200ed.after.kyoscan-$/i]"
}
```

**Reply by darrinsmart** [reply]
I am able to trigger this with macOS Mail.app too. Simply take a screenshot, choose to share via email, and send to an address that is rspamd-enabled.

```
--Apple-Mail=_EE18A1BA-69CA-49EF-B1B6-99A1416703A4
Content-Disposition: inline;
	filename*=utf-8''Image%205%2D6%2D24%20at%2012.42%E2%80%AFPM.jpg
Content-Type: image/jpeg;
	name="=?utf-8?Q?Image_5-6-24_at_12=2E42=E2=80=AFPM=2Ejpg?="
Content-Transfer-Encoding: base64
```

```
X-Spamd-Result: default: False [7.38 / 15.00];
	MIME_BAD_UNICODE(8.00)[0x202fd after Image 5-6-24 at 12.42];
```

**Reply by dragoangel** [reply]
> You are right. I made a wrong assumption. The MIME_BAD_UNICODE is specifically designed to catch such things. [util.has_obscured_unicode(str)](https://rspamd.com/doc/lua/rspamd_util.html#f53782)
> 
> Looking through my notes I found some Kyocera scanners may insert `0x200ed` in filenames.
> 
> I even had a composite for it:
> 
> ```
> LOCAL_MIME_BAD_UNICODE_KYOSCAN {
>     expression = "MIME_BAD_UNICODE[/^0x200ed.after.kyoscan-$/i]"
> }
> ```

BTW, is there any chance that Rspamd doing FP on [Left-to-Right Mark 200E](https://www.compart.com/en/unicode/U+200E)?

**Comment by dragoangel** [comment]
@moisseev it's really the case with MacOS clients... Maybe better put it in default excludes?
@vstakhov what you think?

**Comment by dragoangel** [comment]
created PR https://github.com/rspamd/rspamd/pull/5030 to solve this issue

**Reply by dragoangel** [reply]
@ManChicken1911 jfyi, it's merged to master :)

**Comment by CyberCr33p** [comment]
Looks like this issue is not resolved:

`MIME_BAD_UNICODE • Filename with known obscured unicode characters (8) [0x202fd after Screenshot 2023-10-03 at 1.12.07]`

I use Rspamd 3.11.0 which includes this patch.

**Reply by dragoangel** [reply]
@CyberCr33p if you look at scores which was assigned to such symbol - you will see it 2 instead of 8 as was before. Basically main score assigned to composite, instead of this symbol.

**Comment by CyberCr33p** [comment]
@dragoangel `MIME_BAD_UNICODE` symbol:

<img width="1428" alt="Screenshot 2025-05-16 at 11 56 11" src="https://github.com/user-attachments/assets/63a23b8e-98c7-4cbd-88c3-c8b807f6909e" />

If I’ve got this right, `MIME_BAD_UNICODE` shouldn’t apply here, and `MIME_BAD_EXT_WITH_BAD_UNICODE` only fires when the file extension itself has bad extension - which doesn’t happen for a `.png`

Any idea why is `MIME_BAD_UNICODE` triggered?

**Comment by CyberCr33p** [comment]
I found that MIME_BAD_UNICODE had score 8.00 because it was saved in /var/lib/rspamd/rspamd_dynamic , I removed this file and it changed the score to the default 2.00.

**Reply by dragoangel** [reply]
It's in your conf it's 8.0, not in rspamd default...

**Reply by CyberCr33p** [reply]
Thank you for your responses; they helped me resolve the issue.
